### PR TITLE
fix crash: keep a copy of the return value of setlocale(), it can become invalid

### PR DESCRIPTION
**Pull request by rainers**
spurious crashes occur on Win64, e.g. https://github.com/dlang/dmd/pull/11865

introduced by https://github.com/dlang/phobos/commit/4b416afe1656607ac4f7a1cb704dbc47157d7398

### CODE REVIEW

**Review by CyberShadow — APPROVED** [review]

**Comment by thewilsonator** [comment]
```
grep -nrE '[[:alnum:]][.][.][[:alnum:]]\|[[:alnum:]] [.][.][[:alnum:]]\|[[:alnum:]][.][.] [[:alnum:]]' $(find etc std -name '*.d' \| grep -vE 'std/string.d\|std/uni/package.d') ; test $? -eq 1
--
  | std/stdio.d:3685:            p = p[0..strlen(p)+1].idup.ptr;
```

**Comment by thewilsonator** [comment]
removed auto-merge, style check failed.

File: std/stdio.d [hunk 1]
```diff
@@ -3679,7 +3680,10 @@ void main()
     auto deleteme = testFilename();
     scope(exit) std.file.remove(deleteme);
     const char* oldCt = () @trusted {
-        return setlocale(LC_CTYPE, null);
+        const(char)* p = setlocale(LC_CTYPE, null);
+        if (p) // make a copy, the original might get invalid in subsequent calls to setlocale
+            p = p[0..strlen(p)+1].idup.ptr;
```
Comment (CyberShadow) [inline std/stdio.d:3685]:
```suggestion
            p = p[0 .. strlen(p) + 1].idup.ptr;
```

**Review by WalterBright — DISMISSED** [review]
This won't work with -betterC. Use malloc() instead.

**Comment by CyberShadow** [comment]
@WalterBright The change is in a unittest block

**Comment by Geod24** [comment]
Sorry for the race condition @CyberShadow !

**Pull request merged**